File: app/src/main/java/me/devsaki/hentoid/fragments/library/ErrorsDialogFragment.java
package me.devsaki.hentoid.fragments.library;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.documentfile.provider.DocumentFile;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;

import java.util.ArrayList;
import java.util.List;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.database.CollectionDAO;
import me.devsaki.hentoid.database.ObjectBoxDAO;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.database.domains.ErrorRecord;
import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.enums.StatusContent;
import me.devsaki.hentoid.util.FileHelper;
import me.devsaki.hentoid.util.LogHelper;
import me.devsaki.hentoid.util.ToastHelper;

/**
 * Created by Robb on 11/2018
 * Info dialog for download errors details
 */
public class ErrorsDialogFragment extends DialogFragment {

    private static final String ID = "ID";

    private Parent parent;
    private View rootView;

    public static void invoke(Fragment parent, long id) {
        ErrorsDialogFragment fragment = new ErrorsDialogFragment();

        Bundle args = new Bundle();
        args.putLong(ID, id);
        fragment.setArguments(args);

        fragment.show(parent.getChildFragmentManager(), null);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        parent = (Parent) getParentFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedState) {
        return inflater.inflate(R.layout.dialog_library_errors, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (null == getArguments()) throw new IllegalArgumentException("No arguments found");
        long id = getArguments().getLong(ID, 0);
        if (0 == id) throw new IllegalArgumentException("No ID found");

        Content content;
        CollectionDAO dao = new ObjectBoxDAO(getContext());
        try {
            content = dao.selectContent(id);
        } finally {
            dao.cleanup();
        }
        if (null == content) throw new IllegalArgumentException("Content not found for ID " + id);

        rootView = view;
        updateStats(content);

        View redownloadButton = view.findViewById(R.id.redownload_btn);
        redownloadButton.setOnClickListener(v -> redownload(content));

        View openLogButton = view.findViewById(R.id.open_log_btn);
        openLogButton.setOnClickListener(v -> showErrorLog(content));

        View shareLogButton = view.findViewById(R.id.share_log_btn);
        shareLogButton.setOnClickListener(v -> shareErrorLog(content));
    }

    private void updateStats(@NonNull final Content content) {
        int images;
        int imgErrors = 0;

        Context context = getContext();
        if (null == context) return;
        if (null == content.getImageFiles()) return;

        images = content.getImageFiles().size() - 1; // Don't count the cover

        for (ImageFile imgFile : content.getImageFiles())
            if (imgFile.getStatus() == StatusContent.ERROR) imgErrors++;

        if (0 == images) {
            images = content.getQtyPages();
            imgErrors = images;
        }

        TextView details = rootView.findViewById(R.id.redownload_detail);
        details.setText(context.getString(R.string.redownload_dialog_message, images, images - imgErrors, imgErrors));

        if (content.getErrorLog() != null && !content.getErrorLog().isEmpty()) {
            TextView firstErrorTxt = rootView.findViewById(R.id.redownload_detail_first_error);
            ErrorRecord firstError = content.getErrorLog().get(0);
            String message = context.getString(R.string.redownload_first_error, firstError.getType().getName());
            if (!firstError.getDescription().isEmpty())
                message += String.format(" - %s", firstError.getDescription());
            firstErrorTxt.setText(message);
            firstErrorTxt.setVisibility(View.VISIBLE);
        }
    }

    private LogHelper.LogInfo createLog(@NonNull final Content content) {
        List<LogHelper.LogEntry> log = new ArrayList<>();

        LogHelper.LogInfo errorLogInfo = new LogHelper.LogInfo();
        errorLogInfo.setLogName("Error");
        errorLogInfo.setFileName("error_log" + content.getId());
        errorLogInfo.setNoDataMessage("No error detected.");
        errorLogInfo.setEntries(log);

        List<ErrorRecord> errorLog = content.getErrorLog();
        if (errorLog != null) {
            errorLogInfo.setHeader("Error log for " + content.getTitle() + " [" + content.getUniqueSiteId() + "@" + content.getSite().getDescription() + "] : " + errorLog.size() + " errors");
            for (ErrorRecord e : errorLog)
                log.add(new LogHelper.LogEntry(e.getTimestamp(), e.toString()));
        }

        return errorLogInfo;
    }

    private void showErrorLog(@NonNull final Content content) {
        ToastHelper.toast(R.string.redownload_generating_log_file);

        LogHelper.LogInfo logInfo = createLog(content);
        DocumentFile logFile = LogHelper.writeLog(requireContext(), logInfo);
        if (logFile != null) FileHelper.openFile(requireContext(), logFile);
    }

    private void shareErrorLog(@NonNull final Content content) {
        LogHelper.LogInfo logInfo = createLog(content);
        DocumentFile logFile = LogHelper.writeLog(requireContext(), logInfo);
        if (logFile != null)
            FileHelper.shareFile(requireContext(), logFile.getUri(), "Error log for book ID " + content.getUniqueSiteId());
    }

    private void redownload(@NonNull final Content content) {
        parent.redownloadContent(content);
        dismiss();
    }

    public interface Parent {
        void redownloadContent(final Content content);
    }
}

File: app/src/main/java/me/devsaki/hentoid/fragments/queue/ErrorsDialogFragment.kt
package me.devsaki.hentoid.fragments.queue

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import me.devsaki.hentoid.R
import me.devsaki.hentoid.database.CollectionDAO
import me.devsaki.hentoid.database.ObjectBoxDAO
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.database.domains.ErrorRecord
import me.devsaki.hentoid.databinding.DialogLibraryErrorsBinding
import me.devsaki.hentoid.enums.StatusContent
import me.devsaki.hentoid.fragments.BaseDialogFragment
import me.devsaki.hentoid.util.LogEntry
import me.devsaki.hentoid.util.LogInfo
import me.devsaki.hentoid.util.file.FileHelper
import me.devsaki.hentoid.util.toast
import me.devsaki.hentoid.util.writeLog

/**
 * Info dialog for download errors details
 */
class ErrorsDialogFragment : BaseDialogFragment<ErrorsDialogFragment.Parent>() {

    companion object {
        const val ID = "ID"

        fun invoke(parentFragment: Fragment, id: Long) {
            val args = Bundle()
            args.putLong(ID, id)
            invoke(parentFragment, ErrorsDialogFragment(), args)
        }
    }

    private var binding: DialogLibraryErrorsBinding? = null


    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedState: Bundle?
    ): View {
        binding = DialogLibraryErrorsBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(rootView: View, savedInstanceState: Bundle?) {
        super.onViewCreated(rootView, savedInstanceState)

        requireNotNull(arguments) { "No arguments found" }
        val id = requireArguments().getLong(ID, 0)
        require(0L != id) { "No ID found" }

        val content: Content
        val dao: CollectionDAO = ObjectBoxDAO(requireContext())
        try {
            content = dao.selectContent(id)!!
        } finally {
            dao.cleanup()
        }

        updateStats(content)

        binding?.apply {
            redownloadBtn.setOnClickListener { redownload(content) }
            openLogBtn.setOnClickListener { showErrorLog(content) }
            shareLogBtn.setOnClickListener { shareErrorLog(content) }
        }
    }

    private fun updateStats(content: Content) {
        var images = 0
        var imgErrors = 0
        val context = context ?: return
        content.imageFiles?.let {
            images = it.size - 1 // Don't count the cover
            for (imgFile in it) if (imgFile.status == StatusContent.ERROR) imgErrors++
            if (0 == images) {
                images = content.qtyPages
                imgErrors = images
            }
        }
        binding?.let {
            it.redownloadDetail.text = context.getString(
                R.string.redownload_dialog_message, images, images - imgErrors, imgErrors
            )
            content.errorLog?.let { log ->
                if (!log.isEmpty()) {
                    val firstError = log[0]
                    var message = context.getString(
                        R.string.redownload_first_error,
                        context.getString(firstError.type.displayName)
                    )
                    if (firstError.description.isNotEmpty()) message += String.format(
                        " - %s", firstError.description
                    )
                    it.redownloadDetailFirstError.text = message
                    it.redownloadDetailFirstError.visibility = View.VISIBLE
                }
            }
        }
    }

    private fun createLog(content: Content): LogInfo {
        val log: MutableList<LogEntry> = ArrayList()
        val errorLogInfo = LogInfo("error_log" + content.id)
        errorLogInfo.setHeaderName("Error")
        errorLogInfo.setNoDataMessage("No error detected.")
        errorLogInfo.setEntries(log)
        val errorLog: List<ErrorRecord>? = content.errorLog
        if (errorLog != null) {
            errorLogInfo.setHeader("Error log for " + content.title + " [" + content.uniqueSiteId + "@" + content.site.description + "] : " + errorLog.size + " errors")
            for (e in errorLog) log.add(LogEntry(e.timestamp, e.toString()))
        }
        return errorLogInfo
    }

    private fun showErrorLog(content: Content) {
        toast(R.string.redownload_generating_log_file)
        val logInfo = createLog(content)
        val logFile = requireContext().writeLog(logInfo)
        if (logFile != null) FileHelper.openFile(requireContext(), logFile)
    }

    private fun shareErrorLog(content: Content) {
        val logInfo = createLog(content)
        val logFile = requireContext().writeLog(logInfo)
        if (logFile != null) FileHelper.shareFile(
            requireContext(), logFile.uri, "Error log for book ID " + content.uniqueSiteId
        )
    }

    private fun redownload(content: Content) {
        parent?.redownloadContent(content)
        dismiss()
    }

    interface Parent {
        fun redownloadContent(content: Content)
    }
}